File: admin-server.js
const express = require('express');
const fs = require('fs');
const path = require('path');
const session = require('express-session');
const crypto = require('crypto');
const rateLimit = require('express-rate-limit');

const app = express();
const PORT = process.env.PORT || 5501;

const ADMIN_PASSWORD = process.env.ADMIN_PASSWORD || 'pokgev-4Pimho-wixjep';

const PROJECT_ROOT = path.resolve(__dirname, '..');

const limiter = rateLimit({
    windowMs: 15 * 60 * 1000,
    max: 10,
    message: 'Too many login attempts, please try again later.'
});

app.use(session({
    secret: crypto.randomBytes(64).toString('hex'),
    resave: false,
    saveUninitialized: false,
    cookie: { 
        secure: false, 
        maxAge: 3600000,
        httpOnly: true,
        sameSite: 'strict'
    }
}));

app.use(express.urlencoded({ extended: true }));
app.use(express.json());

app.use(express.static(__dirname));

app.use('/admin/api/auth.php', limiter);

const requireAuth = (req, res, next) => {
    if (req.session && req.session.authenticated) {
        next();
    } else {
        res.status(401).json({ error: 'Not authenticated' });
    }
};

const pages = {
    general: [
        { path: 'index.html', name: 'Homepage' },
        { path: 'about.html', name: 'About' },
        { path: 'contact.html', name: 'Contact' },
        { path: 'research.html', name: 'Research' },
        { path: 'teaching.html', name: 'Teaching' },
        { path: 'notes.html', name: 'Notes' },
        { path: 'blog.html', name: 'Blog' },
        { path: 'Track.html', name: 'Track and Field' }
    ],
    courses: [
        { path: 'courses/precalculus-11.html', name: 'Precalculus 11' },
        { path: 'courses/precalculus-12.html', name: 'Precalculus 12' },
        { path: 'courses/calculus-1.html', name: 'Calculus I' },
        { path: 'courses/calculus-2.html', name: 'Calculus II' },
        { path: 'courses/calculus-3.html', name: 'Calculus III' },
        { path: 'courses/real-analysis.html', name: 'Real Analysis' },
        { path: 'courses/introduction-to-proofs.html', name: 'Introduction to Proofs' },
        { path: 'courses/number-theory.html', name: 'Elementary Number Theory' },
        { path: 'courses/algebra-I.html', name: 'Algebra I' },
        { path: 'courses/algebra-II.html', name: 'Algebra II' },
        { path: 'courses/linear-algebra.html', name: 'Linear Algebra' }
    ],
    olympiad: [
        { path: 'olympiad.html', name: 'Olympiad' },
        { path: 'undergraduate.html', name: 'Undergraduate' },
        { path: 'graduate.html', name: 'Graduate' }
    ]
};

app.post('/admin/api/auth.php', (req, res) => {
    const { username, password } = req.body;
    
    if (username === 'admin' && password === ADMIN_PASSWORD) {
        req.session.authenticated = true;
        req.session.username = username;
        req.session.csrf_token = crypto.randomBytes(32).toString('hex');
        req.session.loginTime = Date.now();
        res.json({ success: true, csrf_token: req.session.csrf_token });
    } else {
        res.status(401).json({ success: false, error: 'Invalid credentials' });
    }
});

app.get('/admin/api/auth.php', requireAuth, (req, res) => {
    res.json({ 
        authenticated: true, 
        username: req.session.username, 
        csrf_token: req.session.csrf_token,
        loginTime: req.session.loginTime
    });
});

app.get('/admin/api/content.php', requireAuth, (req, res) => {
    const action = req.query.action;
    
    if (action === 'list') {
        res.json({ pages, sections: Object.keys(pages) });
        return;
    }
    
    if (action === 'content' && req.query.page) {
        const page = req.query.page;
        const filepath = path.join(PROJECT_ROOT, page);
        
        if (fs.existsSync(filepath)) {
            const content = fs.readFileSync(filepath, 'utf8');
            
            const backupDir = path.join(PROJECT_ROOT, 'admin', 'backups', new Date().toISOString().split('T')[0]);
            if (!fs.existsSync(backupDir)) {
                fs.mkdirSync(backupDir, { recursive: true });
            }
            
            const backupFile = path.join(backupDir, page.replace(/\//g, '_') + '.' + Date.now());
            fs.writeFileSync(backupFile, content);
            
            res.json({ content });
        } else {
            res.status(404).json({ error: 'Page not found' });
        }
        return;
    }
    
    res.status(400).json({ error: 'Invalid action' });
});

app.post('/admin/api/content.php', requireAuth, (req, res) => {
    const { page, content, csrf_token } = req.body;
    
    if (!page || !content) {
        return res.status(400).json({ error: 'Missing required fields' });
    }
    
    if (csrf_token !== req.session.csrf_token) {
        return res.status(403).json({ error: 'Invalid CSRF token' });
    }
    
    if (page.includes('..') || !page.endsWith('.html')) {
        return res.status(403).json({ error: 'Invalid page path' });
    }
    
    const filepath = path.join(PROJECT_ROOT, page);
    
    try {
        const backupDir = path.join(PROJECT_ROOT, 'admin', 'backups', new Date().toISOString().split('T')[0]);
        if (!fs.existsSync(backupDir)) {
            fs.mkdirSync(backupDir, { recursive: true });
        }
        
        if (fs.existsSync(filepath)) {
            fs.writeFileSync(path.join(backupDir, page.replace(/\//g, '_') + '.' + Date.now()), fs.readFileSync(filepath));
        }
        
        fs.writeFileSync(filepath, content);
        res.json({ success: true });
    } catch (error) {
        res.status(500).json({ error: 'Failed to save: ' + error.message });
    }
});

app.get('/admin/api/history.php', requireAuth, (req, res) => {
    const page = req.query.page;
    const historyDir = path.join(PROJECT_ROOT, 'admin', 'backups');
    
    if (!fs.existsSync(historyDir)) {
        return res.json({ versions: [] });
    }
    
    const versions = [];
    const dateDirs = fs.readdirSync(historyDir).filter(d => !d.startsWith('.'));
    
    dateDirs.forEach(date => {
        const dateDir = path.join(historyDir, date);
        if (fs.statSync(dateDir).isDirectory()) {
            const files = fs.readdirSync(dateDir);
            files.forEach(file => {
                if (file.includes(page.replace(/\//g, '_'))) {
                    versions.push({
                        date: date,
                        file: file,
                        path: path.join(dateDir, file)
                    });
                }
            });
        }
    });
    
    res.json({ versions: versions.sort((a, b) => b.date.localeCompare(a.date)) });
});

app.post('/admin/api/logout.php', (req, res) => {
    req.session.destroy(() => {
        res.json({ success: true });
    });
});

app.listen(PORT, () => {
    console.log(`Admin server running at http://localhost:${PORT}`);
    console.log(`Visit http://localhost:${PORT}/admin/ to access the editor`);
});

File: admin-server.py
#!/usr/bin/env python3
import http.server
import socketserver
import json
import os
import urllib.parse
from http import cookies
import hashlib
import time
import shutil

PORT = 5502
BASE_DIR = os.path.dirname(os.path.abspath(__file__))
ADMIN_DIR = os.path.join(BASE_DIR, 'admin')

sessions = {}

def get_session_id():
    return hashlib.sha256(str(time.time()).encode()).hexdigest()[:16]

class AdminHandler(http.server.SimpleHTTPRequestHandler):
    def do_GET(self):
        parsed = urllib.parse.urlparse(self.path)
        path = parsed.path
        query = dict(urllib.parse.parse_qsl(parsed.query))
        
        if path == '/admin/api/auth.php':
            self.send_response(401)
            self.send_header('Content-Type', 'application/json')
            self.end_headers()
            self.wfile.write(json.dumps({'success': False, 'error': 'Not authenticated'}).encode())
            return
        
        if path == '/admin/api/content.php':
            action = query.get('action', '')
            
            if action == 'list':
                self.send_response(200)
                self.send_header('Content-Type', 'application/json')
                self.end_headers()
                pages = {
                    'index.html': 'Homepage',
                    'teaching.html': 'Teaching',
                    'courses/calculus-1.html': 'Calculus I',
                    'courses/calculus-2.html': 'Calculus II',
                    'courses/calculus-3.html': 'Calculus III',
                    'courses/linear-algebra.html': 'Linear Algebra',
                    'courses/real-analysis.html': 'Real Analysis',
                    'courses/introduction-to-proofs.html': 'Introduction to Proofs',
                    'courses/number-theory.html': 'Elementary Number Theory',
                    'courses/algebra-I.html': 'Algebra I',
                    'courses/algebra-II.html': 'Algebra II',
                    'research.html': 'Research',
                    'about.html': 'About',
                    'contact.html': 'Contact',
                    'notes.html': 'Notes',
                    'blog.html': 'Blog',
                    'courses/lecture-notes-1.html': 'Lecture Notes 1',
                    'courses/lecture-notes-2.html': 'Lecture Notes 2',
                    'courses/lecture-notes-3.html': 'Lecture Notes 3',
                    'courses/lecture-notes-4.html': 'Lecture Notes 4',
                    'courses/lecture-notes-5.html': 'Lecture Notes 5',
                    'courses/lecture-notes-6.html': 'Lecture Notes 6',
                    'courses/lecture-notes-7.html': 'Lecture Notes 7',
                    'courses/lecture-notes-8.html': 'Lecture Notes 8',
                    'courses/lecture-notes-9.html': 'Lecture Notes 9',
                    'courses/lecture-notes-10.html': 'Lecture Notes 10',
                    'courses/lecture-notes-11.html': 'Lecture Notes 11',
                    'courses/lecture-notes-12.html': 'Lecture Notes 12',
                    'courses/lecture-notes-13.html': 'Lecture Notes 13',
                    'courses/lecture-notes-14.html': 'Lecture Notes 14'
                }
                self.wfile.write(json.dumps({'pages': pages}).encode())
                return
            
            if action == 'content':
                page = query.get('page', '')
                filepath = os.path.join(ADMIN_DIR, page)
                if os.path.exists(filepath):
                    with open(filepath, 'r') as f:
                        content = f.read()
                    self.send_response(200)
                    self.send_header('Content-Type', 'application/json')
                    self.end_headers()
                    self.wfile.write(json.dumps({'content': content}).encode())
                    return
                else:
                    self.send_response(404)
                    self.send_header('Content-Type', 'application/json')
                    self.end_headers()
                    self.wfile.write(json.dumps({'error': 'Page not found'}).encode())
                    return
        
        super().do_GET()
    
    def do_POST(self):
        parsed = urllib.parse.urlparse(self.path)
        path = parsed.path
        
        content_length = int(self.headers.get('Content-Length', 0))
        post_data = self.rfile.read(content_length).decode('utf-8')
        
        if path == '/admin/api/auth.php':
            data = dict(urllib.parse.parse_qsl(post_data))
            username = data.get('username', '')
            password = data.get('password', '')
            
            if username == 'admin' and password == 'pokgev-4Pimho-wixjep':
                session_id = get_session_id()
                sessions[session_id] = {'username': username, 'csrf_token': get_session_id()}
                
                self.send_response(200)
                self.send_header('Content-Type', 'application/json')
                self.send_header('Set-Cookie', f'session_id={session_id}; Path=/')
                self.end_headers()
                self.wfile.write(json.dumps({'success': True, 'csrf_token': sessions[session_id]['csrf_token']}).encode())
                return
            else:
                self.send_response(401)
                self.send_header('Content-Type', 'application/json')
                self.end_headers()
                self.wfile.write(json.dumps({'success': False, 'error': 'Invalid credentials'}).encode())
                return
        
        if path == '/admin/api/content.php':
            data = dict(urllib.parse.parse_qsl(post_data))
            page = data.get('page', '')
            content = data.get('content', '')
            csrf_token = data.get('csrf_token', '')
            
            if page and content:
                filepath = os.path.join(ADMIN_DIR, page)
                if not os.path.exists(filepath):
                    self.send_response(404)
                    self.send_header('Content-Type', 'application/json')
                    self.end_headers()
                    self.wfile.write(json.dumps({'error': 'Page not found'}).encode())
                    return
                
                backup_dir = os.path.join(ADMIN_DIR, 'backups', time.strftime('%Y-%m-%d'))
                os.makedirs(backup_dir, exist_ok=True)
                if os.path.exists(filepath):
                    shutil.copy(filepath, os.path.join(backup_dir, os.path.basename(page)))
                
                with open(filepath, 'w') as f:
                    f.write(content)
                
                self.send_response(200)
                self.send_header('Content-Type', 'application/json')
                self.end_headers()
                self.wfile.write(json.dumps({'success': True}).encode())
                return
        
        super().do_POST()

with socketserver.TCPServer(("", PORT), AdminHandler) as httpd:
    print(f"Serving at port {PORT}")
    print(f"Visit http://localhost:{PORT}/admin/")
    httpd.serve_forever()